New Permutations package

TheGrateSalmon opened the pull request:
New package implementing a `Permutation` type by Sean Grate. Its intent is to have a standard and consistent type for permutations as more combinatorics-focused packages get added to Macaulay2. Much of the inspiration for the implemented methods came from the [Sage implementation](https://doc.sagemath.org/html/en/reference/combinat/sage/combinat/permutation.html). This also resolves [Issue #2581](https://github.com/Macaulay2/M2/issues/2581).

mahrud commented:
1. Add your package to the `=distributed-packages` file, otherwise it's basically ignored.
2. I think having both `Permutations.m2` and `Permutations/Permutations.m2` could lead to hard to decipher errors (e.g. try `needsPackage "Permutations"` inside the `Permutations` directory). My suggestion would be to put most of the content of  `Permutations/Permutations.m2` in `Permutations.m2`, and name the other two files `tests.m2` and `docs.m2`.
3. Typos:
```
M2/Macaulay2/packages/Permutations/PermutationsDOC.m2:241: postive ==> positive
M2/Macaulay2/packages/Permutations/PermutationsDOC.m2:622: simulataneously ==> simultaneously
```

d-torrance commented on M2/Macaulay2/packages/Permutations.m2:
```diff
@@ -0,0 +1,98 @@
+        -- "SimplicialComplexes",
+        -- "SimplicialDecomposability",
+        -- "Posets",
+        -- "MinimalPrimes"
```
Is there a reason all these packages in `PackageExports` are commented out?

TheGrateSalmon replied:
No, a byproduct of copying from MatrixSchubert I believe. Should be fixed in some new commits.

d-torrance commented on M2/Macaulay2/packages/Permutations.m2:
```diff
@@ -0,0 +1,98 @@
+        -- "Posets",
+        -- "MinimalPrimes"
+    },
+    DebuggingMode => true
```
`DebuggingMode` should either be set to `false`, or this option removed completely (since `false` is the default).

TheGrateSalmon replied:
This should be corrected in some new commits.

d-torrance commented on M2/Macaulay2/tests/normal/permutations.m2:
```diff
@@ -1,5 +0,0 @@
-v = {1,2,4,3,0}
-w = {3,1,2,4,0}
-assert(f_(v_w)==(f_v)_w)
-assert(f_(v_w)==(f_v)*(f_w))
```
Why is this test file being removed?

mahrud replied:
It's moved to the package tests.

TheGrateSalmon replied:
Originally, I had tried to implement all of this directly into M2, i.e., not as a package, so I had written all of its tests here. Upon recommendation, I then tried to refactor all of permutations code I had written into a package and I simply forgot that M2 actually had this file to begin with, so it got deleted during the refactor.

d-torrance replied:
Oh, ok -- I see it in the package tests now.  But it doesn't use anything from the package, so should it stay where it was?

TheGrateSalmon replied:
I think it staying where it was makes sense as it uses M2's base functionality.

mahrud commented on M2/Macaulay2/packages/Permutations/Permutations.m2:
```diff
@@ -0,0 +1,311 @@
+Permutation = new Type of BasicList
+Permutation.synonym = "permutation"
+
+isValidPermutation = method(TypicalValue => Boolean)
+isValidPermutation List := Boolean => w -> (((sort w) == toList(1..#w)) and not (#w == 0))
+
+new Permutation from BasicList := (typeofPermutation,w) -> (
+    if not isValidPermutation(w) then error(toString w | " is not a valid permutation in one-line notation.");
+    w
+)
```
A better approach is this: don't check for correctness in the constructor, because it will slow down people's code (e.g. if they're making lots of very large permutations, and they know their permutations are definitely valid, then there's no need to check); instead change your verification method to `isWellDefined Permutation`, so people can verify that the permutation they constructed is correct.

Also, I think you might want `VisibleList` instead of `BasicList` in both places?

TheGrateSalmon replied:
That makes sense and should be corrected in some new commits. Related, but can I also use `VisibleList` as a type anywhere I would expect a `List` and `Sequence` to behave identically? I would hope that this would mean some code could be condensed and reduce the total number of overloads for a method.

mahrud replied:
Yes, List, Sequence, Array, and AngleBarList inherit from VisibleList. On the other hands, loads of things inherit from BasicList (think Option, Command, URL, etc.)

mahrud commented on M2/Macaulay2/packages/Permutations/Permutations.m2:
```diff
@@ -0,0 +1,311 @@
+)
+
+permutation = method(TypicalValue => Permutation)
+permutation List := Permutation => w -> new Permutation from w
```
Once you declared `TypicalValue => Permutation`, you don't need to list `Permutation => ` in front of every method.

(On the other hand, you should think about whether anyone else might want to use the method name `permutation` for a different purpose and have it return a different object, in which case fixing `TypicalValue => Permutation` would make their life harder)

TheGrateSalmon replied:
I didn't think about that side effect of `TypicalValue`, so it has been more carefully thought out and corrected in some new commits.

TheGrateSalmon commented:
I should also mention that this package includes code for pattern avoidance and is taken directly from the `MatrixSchubert` package. The idea is that since pattern avoidance is inherent to permutations, it'll go into `Permutations`, and following the adoption of this package, the `MatrixSchubert` package might be overhauled to use the `Permutation` type from this package for consistency.

mahrud commented on M2/Macaulay2/packages/Permutations.m2:
```diff
@@ -0,0 +1,383 @@
+-- Pattern avoidance
+------------------------------------
+avoidsPattern = method(TypicalValue => Boolean)
+avoidsPattern (Permutation,List) := (w, pattern) -> (
```
You should mention here that this code is borrowed from MatrixSchubert's `isPatternAvoiding` (and any others that may be in similar situation), and also get in touch with their authors to see how the two packages should interoperate (though that doesn't have to happen in this PR).

TheGrateSalmon replied:
Ah, I didn't think to mention it in the code itself because of my comment above.

> I should also mention that this package includes code for pattern avoidance and is taken directly from the `MatrixSchubert` package. The idea is that since pattern avoidance is inherent to permutations, it'll go into `Permutations`, and following the adoption of this package, the `MatrixSchubert` package might be overhauled to use the `Permutation` type from this package for consistency.

As one of the authors myself, I will be coordinating with Ayah to manage the overhaul for `MatrixSchubert`.

mahrud replied:
Oh I didn't realize you are one of the authors. Never mind then haha.

mahrud commented on M2/Macaulay2/packages/Permutations/docs.m2:
```diff
@@ -0,0 +1,1296 @@
+-- avoidsPatterns
+doc ///
+  Key
+    avoidsPatterns
```
Under `SeeAlso`, you can link to `"MatrixSchubert::avoidsAllPatterns"` and `"MatrixSchubert::isPatternAvoiding"`.

mahrud commented on M2/Macaulay2/packages/Permutations.m2:
```diff
@@ -0,0 +1,383 @@
+new Permutation from VisibleList := (typeofPermutation,w) -> w
+
+permutation = method()
+permutation List := Permutation => w -> new Permutation from w
```
If you want, I think it would be reasonable to move `permutations`, `inversePermutation`, and (possibly?) `uniquePermutations` into this package (along with their tests and documentation), and comment that they were moved from Core. This may break some tests in other packages, in which case we will add this package as their dependency.

TheGrateSalmon replied:
I think this make sense and I'll look into it.

mahrud commented on M2/Macaulay2/packages/Permutations.m2:
```diff
@@ -0,0 +1,383 @@
+------------------------------------
+-- Local utilities
+------------------------------------
+to1Index := w -> (w / (i -> i+1))
+to0Index := w -> (w / (i -> i-1))
```
This is fine, but I'm curious if there was a problem with my suggestion about prepending 0 to switch to 0-index: https://github.com/Macaulay2/M2/issues/2581#issuecomment-2148157938

TheGrateSalmon replied:
No, those helper functions seemed like the most obvious thing to do at the time I had written the bulk of this code (December 2023).

pzinn commented:
One incompatibility with the built-in permutations is that they start at 0, not 1.

mahrud commented:
> One incompatibility with the built-in permutations is that they start at 0, not 1.

Yes, although MatrixSchubert is also 1-indexed. I realize I'm not the main audience of this, so I don't have a strong opinion. Would it be much harder to check whether a permutation is 0 or 1 indexed in every algorithm?

TheGrateSalmon commented:
> > One incompatibility with the built-in permutations is that they start at 0, not 1.
> 
> Yes, although MatrixSchubert is also 1-indexed. I realize I'm not the main audience of this, so I don't have a strong opinion. Would it be much harder to check whether a permutation is 0 or 1 indexed in every algorithm? Or maybe have easy to use functions for going forth and back?

I think a good solution would be some sort of configuration option when loading the package. So, it defaults to $1$-indexing, but the user can switch to $0$-indexing if they prefer. Something like `needsPackage("Permutations", ZeroIndex=>true)`.

d-torrance commented:
@TheGrateSalmon, @mahrud: What's the status of this PR?  Should we include this package in the release as it is, or is still waiting from some changes?

TheGrateSalmon commented:
> 1. `VisibleList_Permutation` (and `MutableList`), `Matrix_Permutation` and `Matrix^Permutation` (and `MutableMatrix`) for permuting elements of lists and columns or rows of matrices.

This is somewhat implemented in the package, though with a different syntax. I think your suggestion makes a bit more sense. 

> 2. A comparison operator `Permutation ? Permutation` based on the Bruhat order (is this available elsewhere in M2?)
> 
> 3. A method to get the poset of Bruhat order and the Hasse diagram of S_n (including maybe a way to get the subposet below or above a given permutation)

These  items would be wonderful to have. I'm not sure what the best way to display the Hasse diagram would be, however. I don't think I will get around to implementing them in the current release though.

> 4. `matrix Permutation` to get the permutation matrix (I know `permToMatrix` in `MatrixSchubert` does this, but given a `Permutation` type you could just define `matrix` on it)

This also exists as `toMatrix`, but this syntax makes a bit more sense as well, so it should be an easy change.



> @TheGrateSalmon, @mahrud: What's the status of this PR? Should we include this package in the release as it is, or is still waiting for some changes?

I should be able to incorporate some of @mahrud's suggestions this weekend. After that, from my understanding, it sounds like this package will be ready for its initial release.

d-torrance commented:
> These items would be wonderful to have. I'm not sure what the best way to display the Hasse diagram would be, however. I don't think I will get around to implementing them in the current release though.

There's a `hasseDiagram` method already in the [`Posets`](https://macaulay2.com/doc/Macaulay2/share/doc/Macaulay2/Posets/html/index.html) package that you could use.

mahrud commented:
> There's a `hasseDiagram` method already in the [`Posets`](https://macaulay2.com/doc/Macaulay2/share/doc/Macaulay2/Posets/html/index.html) package that you could use.

How do I get the poset of $$S_n$$ to begin with?

d-torrance commented:
> How do I get the poset of S n to begin with?

How's the partial order defined?  If it's easy to either list all the pairs $(\sigma,\tau)$ with $\sigma\leq\tau$ or write a function that returns true when $\sigma\leq\tau$, then the [`poset`](https://macaulay2.com/doc/Macaulay2/share/doc/Macaulay2/Posets/html/_poset.html) method should work.

mahrud commented:
$S_n$ has $n!$ elements so listing all ordered pairs _works_ but definitely not the optimal solution. The Hasse diagram has far fewer edges but I don't know how many exactly.

TheGrateSalmon commented:
I've added @mahrud's request for `VisibleList`, though I'm not sure how to do the same for `MutableList` (mainly because I'm not sure what the expected behavior with something like `Bag` should be). I've also changed `toMatrix` to `matrix`.

> S n has n ! elements so listing all ordered pairs _works_ but definitely not the optimal solution. The Hasse diagram has far fewer edges but I don't know how many exactly.

If I am remembering correctly, the Bruhat order gives a _graded_ Hasse diagram with a unique maximal element, the longest word permutation $(n,n-1,\dots,1)$. So, I would think a breadth-first search from that element should be good enough, though this is without thinking about the problem too carefully.

mahrud commented:
This looks good! Could you resolve the conflict so the tests can run?
